[HLSL] Introduce address space `hlsl_constant(2)` for constant buffer declarations

hekota opened the pull request:
Introduces a new address space `hlsl_constant(2)` for constant buffer declarations.

This address space is applied to declarations inside `cbuffer` block. Later on, it will also be applied to `ConstantBuffer<T>` syntax and the default `$Globals` constant buffer.

Clang codegen translates constant buffer declarations to global variables and loads from `hlsl_constant(2)` address space. More work coming soon will include addition of metadata that will map these globals to individual constant buffers and enable their transformation to appropriate constant buffer load intrinsics later on in an LLVM pass.

Fixes #123406

hekota committed:
[HLSL] Add address space `hlsl_constant(2)` for constant buffer declarations

hekota commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -100,22 +100,6 @@ GlobalVariable *replaceBuffer(CGHLSLRuntime::Buffer &Buf) {
       llvm::formatv("{0}{1}", Buf.Name, Buf.IsCBuffer ? ".cb." : ".tb."),
       GlobalValue::NotThreadLocal);
 
-  IRBuilder<> B(CBGV->getContext());
-  Value *ZeroIdx = B.getInt32(0);
-  // Replace Const use with CB use.
-  for (auto &[GV, Offset] : Buf.Constants) {
-    Value *GEP =
-        B.CreateGEP(Buf.LayoutStruct, CBGV, {ZeroIdx, B.getInt32(Offset)});
-
-    assert(Buf.LayoutStruct->getElementType(Offset) == GV->getValueType() &&
-           "constant type mismatch");
-
-    // Replace.
-    GV->replaceAllUsesWith(GEP);
-    // Erase GV.
-    GV->removeDeadConstantUsers();
-    GV->eraseFromParent();
-  }
```
I am removing this part to make sure we preserve the loads in the new address space, but there is more reworking to be done in this area.

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -455,14 +456,22 @@ void createHostLayoutStructForBuffer(Sema &S, HLSLBufferDecl *BufDecl) {
+      QualType NewTy =
+          AST.getAddrSpaceQualType(VD->getType(), LangAS::hlsl_constant);
+      VD->setType(NewTy);
+      VD->setInit(nullptr);
```
Does this `VD->setInit(nullptr);` silently get rid of an initializer if there was one?  This feels a bit sketchy, unless I'm missing something.  I know we don't currently support capturing initializers for constant buffer values, but it is part of HLSL syntax and could in theory be captured.  Silently erasing it from the AST node at this point seems weird.

hekota replied:
If we need to preserve the initializers in the AST even though they are not used, we could mark the generated global variable as `externally_initialized`. The global gets removed in the final DXIL after all.

tex3d commented on clang/test/AST/HLSL/ast-dump-comment-cbuffer-tbuffer.hlsl:
```diff
@@ -43,8 +43,8 @@ tbuffer B {
-// AST-NEXT: VarDecl {{.*}} a 'float'
-// AST-NEXT: VarDecl {{.*}} b 'int'
+// AST-NEXT: VarDecl {{.*}} a 'hlsl_constant float'
+// AST-NEXT: VarDecl {{.*}} b 'hlsl_constant int'
```
Might we want a different address space for tbuffer, potentially?

hekota replied:
Correct. I will update this test to be just for `cbuffer` and will handle `tbuffer` separately.

tex3d commented on clang/test/AST/HLSL/cbuffer.hlsl:
```diff
@@ -39,7 +39,7 @@ typedef float EmptyArrayTypedef[10][0];
 // CHECK: HLSLResourceAttr {{.*}} Implicit CBuffer
 cbuffer CB {
-  // CHECK: VarDecl {{.*}} col:9 used a1 'float'
+  // CHECK: VarDecl {{.*}} col:9 used a1 'hlsl_constant float'
```
Will this also impact type names in diagnostics?

hekota replied:
Probably, the address space is attached to `QualType`.

tex3d reviewed (commented):
I think this looks good, though I had some questions.

Looks like a test update is needed:
```
error: 'expected-error' diagnostics expected but not seen: 
  File /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-q72qr-1/llvm-project/github-pull-requests/clang/test/SemaTemplate/address_space-dependent.cpp Line 46: address space is larger than the maximum supported (8388586)
error: 'expected-error' diagnostics seen but not expected: 
  File /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-q72qr-1/llvm-project/github-pull-requests/clang/test/SemaTemplate/address_space-dependent.cpp Line 46: address space is larger than the maximum supported (8388585)
```

s-perron commented on clang/test/CodeGenHLSL/cbuf.hlsl:
Could we add a test for SPIR-V to make sure it gets the correct address space as well? They have different target info, so the are different code paths.

hekota replied:
I have added SPIR-V variants of the 3 codegen tests.

s-perron reviewed (approved):
LGTM. The SPIR-V code seems correct. It would be useful to have a SPIR-V specific test.

hekota commented on clang/lib/Basic/Targets/DirectX.h:
```diff
@@ -42,6 +42,7 @@ static const unsigned DirectXAddrSpaceMap[] = {
     0, // ptr32_uptr
     0, // ptr64
     3, // hlsl_groupshared
+    2, // hlsl_constant
```
Yes, not updating all of the maps produces a compile error.

hekota commented on clang/lib/Basic/Targets/AMDGPU.cpp:
```diff
@@ -83,6 +83,7 @@ const LangASMap AMDGPUTargetInfo::AMDGPUDefIsPrivMap = {
     llvm::AMDGPUAS::FLAT_ADDRESS, // ptr32_uptr
     llvm::AMDGPUAS::FLAT_ADDRESS, // ptr64
     llvm::AMDGPUAS::FLAT_ADDRESS, // hlsl_groupshared
+    llvm::AMDGPUAS::FLAT_ADDRESS, // hlsl_constant
```
Interesting. Sounds like it is useful to use `static const unsigned[]` then instead of `LangASMap` since it gives you a compile error when the dimension does not match.

hekota committed:
code review feedback
- preserve init and use externally_initialized
- update tests and add SPIRV variant
- reorder switch in TypePrinter.cpp

Keenuts commented on clang/test/CodeGenHLSL/cbuf.hlsl:
```diff
@@ -1,7 +1,14 @@
+// CHECK: @c = external addrspace(2) externally_initialized global float, align 4
+// CHECK: @d = external addrspace(2) externally_initialized global double, align 8
+
 // CHECK: @[[CB:.+]] = external constant { float, double }
```
Why are those still emitted?

Keenuts replied:
Ok, so this `CB` will be replaced by the CBV, and the actual globals will be deleted in favor of the intrinsic if I understand, thanks!

Keenuts commented on clang/test/CodeGenHLSL/static_global_and_function_in_cb.hlsl:
```diff
@@ -1,16 +1,21 @@
+  // CHECK: @a = external addrspace(2) externally_initialized global float, align 4
+  float a;
+  // CHECK: @_ZL1b = internal global float 3.000000e+00, align 4
   static float b = 3;
```
This shouldn't be legal no?
(https://github.com/llvm/llvm-project/issues/117715 )

Keenuts replied:
If the end-goal is not to support this, but we "do" because we just inherit the behavior for now, I'd be in favor of checking that this is disallowed and mark as XFAIL with this issued referenced

hekota replied:
At this point it is not clear what is the end goal. The static decl here tests that is does not get added to the cbuffer layout struct or the new address space. When/if we prohibit static decls in cbuffers this test will surely flare up and will be fixed up.

Keenuts reviewed (approved):
Thanks! Just some questions, otherwise LGTM.
For the AMDGPU address space, shouldn't this be `CONSTANT_ADDRESS` ? (Might want to double check this with Matt?)

hekota committed:
map hlsl_constant to CONSTANT_ADDRESS for AMDGPU

hekota committed:
Update tests after rebase
